### ISSUE TITLE
Numba binning not producing the same result when providing bin edges as list

### ISSUE THREAD

**Issue by rettigl** [OpenCOMPES/sed#66]
**Describe the bug**
When the new binning function get the bin edges as a list, it does not produce the same result as if it gets the bins as number of bins and bin ranges. The version with bin edges as list seems to ignore the last edge (?).

**To Reproduce**
The following code executed in in the example data notebook reproduces the problem:

```
binAxes = ["t"]
nBins = [300]
binRanges = [(65000, 67000)]
res1 = bin_dataframe(
    df=ddf,
    axes=binAxes,
    bins=nBins,
    ranges=binRanges,
    histMode="numba",
    nCores=20,
)
binAxes = ["t"]
res2 = bin_dataframe(
    df=ddf,
    axes=binAxes,
    bins=[(65000, 67000,300)],
    histMode="numba",
    nCores=20,
)
binAxes = ["t"]
nBins = np.linspace(65000, 67000, 300, endpoint=True)
res3 = bin_dataframe(
    df=ddf,
    axes=binAxes,
    bins=nBins,
    histMode="numba",
    nCores=20,
)
res1.plot()
res2.plot()
res3.plot()
print((res1.t==res2.t).all())
print((res1==res2).all())
print((res1.t==res2.t).all())
print((res1==res3).all())
```
Output:
```
<xarray.DataArray 't' ()>
array(True)
<xarray.DataArray ()>
array(True)
<xarray.DataArray 't' ()>
array(True)
<xarray.DataArray ()>
array(False)
```
**Expected behavior**
The results should be identical, as the t-axes are identical

**Screenshots**
Graph:
![output](https://user-images.githubusercontent.com/53396064/162589247-a4e78d61-1127-415c-a165-869d84761a00.png)

**Comment by zainsohail04** [comment]
Is this bug still there?

**Comment by rettigl** [comment]
This bug still exists. Apparently, our tests for binning are very incomplete. In particular, there are not tests for bin_partition or bin_dataframe. 
Indeed, binning with provided bins as arrays fails right now, because the result array size is wrongly calculated. This at least is fixed in PR #98 
Additionally, I encountered that performance in binning with arrays seems to be substantially worse than with bins as int for high-dimensional data, and under certain circumstances binning hangs completely in an infinite loop. I expect this to happen in binsearch, but this needs further investigation.

**Comment by rettigl** [comment]
I've been chasing this for some time now. What I have found so far is that the current version of the numba bin routines for the array case indeed reproduce numpy exactly.
However, the bins version does not, at least for some very edge cases. An example to reproduce is the following (in the bugfix/binning-tuple branch, but this is only important for the exact numbers):
```
import dask
import pandas as pd
import numpy as np
from sed.binning import bin_dataframe
test_ddf = dask.dataframe.from_pandas(pd.DataFrame(np.array([65008.]), columns=["t"]), npartitions=2)
test_ddf
binAxes = ["t"]
nBins = [300]
binRanges = [(65000, 66600)]
res1 = bin_dataframe(
    df=test_ddf,
    axes=binAxes,
    bins=nBins,
    ranges=binRanges,
    hist_mode="numba",
)
res2 = bin_dataframe(
    df=test_ddf,
    axes=binAxes,
    bins=nBins,
    ranges=binRanges,
    hist_mode="numpy",
)
(res1.values == res2.values).all()
```

The reason for this seems to be a numerical instability in calculation of the "delta" step in _hist_from_bins, which applies to every bin step and this affects events very close to bin edges. The case above is nominally exactly on a bin edge, and should go to bin 2, but is placed in bin 2 by _hist_from_bins, because j=1.9999999999995544.
Any ideas on how to fix this and keep the speed benefit are highly welcome.

**Comment by rettigl** [comment]
A possible solution could be to remove the special code for bin_edges, and loop everything through _hist_from_bins. This is slightly less performant, however the differences are minute.
On our machine, binnig of a 4D dataset with corections, jittering etc. (10 M evens, [100, 100, 200, 50] bins) takes
_hist_from_bin_edges: 1m54.8s
_hist_from_bins: 2m1.4s

**Comment by daviddoji** [comment]
> I've been chasing this for some time now. What I have found so far is that the current version of the numba bin routines for the array case indeed reproduce numpy exactly. However, the bins version does not, at least for some very edge cases. An example to reproduce is the following (in the bugfix/binning-tuple branch, but this is only important for the exact numbers):
> 
> ```
> import dask
> import pandas as pd
> import numpy as np
> from sed.binning import bin_dataframe
> test_ddf = dask.dataframe.from_pandas(pd.DataFrame(np.array([65008.]), columns=["t"]), npartitions=2)
> test_ddf
> binAxes = ["t"]
> nBins = [300]
> binRanges = [(65000, 66600)]
> res1 = bin_dataframe(
>     df=test_ddf,
>     axes=binAxes,
>     bins=nBins,
>     ranges=binRanges,
>     hist_mode="numba",
> )
> res2 = bin_dataframe(
>     df=test_ddf,
>     axes=binAxes,
>     bins=nBins,
>     ranges=binRanges,
>     hist_mode="numpy",
> )
> (res1.values == res2.values).all()
> ```
> 
> The reason for this seems to be a numerical instability in calculation of the "delta" step in _hist_from_bins, which applies to every bin step and this affects events very close to bin edges. The case above is nominally exactly on a bin edge, and should go to bin 2, but is placed in bin 2 by _hist_from_bins, because j=1.9999999999995544. Any ideas on how to fix this and keep the speed benefit are highly welcome.

Although this is not a solution, whenever I'm comparing the values of two arrays, I use:
```numpy.testing.assert_allclose(arr1, arr2)``` 
This check avoids floating-point rounding errors up to a certain tolerance that you can specify. 

[Documentation](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose)

**Comment by rettigl** [comment]
> Although this is not a solution, whenever I'm comparing the values of two arrays, I use:
> `numpy.testing.assert_allclose(arr1, arr2)`
> This check avoids floating-point rounding errors up to a certain tolerance that you can specify.

This is what the tests are doing. Off course we could say we don't care about one or the other electron falling off the cliff into onother bin, and say e.g. we only want the two methods to be the same in 99.9% of the cases. That's why I bring it up for discussion here.

A more thorough performance comparison of the test case above yields:
_hist_from_bin_edges: 1min 43s ± 208 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
_hist_from_bins: 1min 50s ± 215 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

i.e. the latter is about 6.8% slower. I can get more examples for different sitatutions.

**Comment by steinnymir** [comment]
If a better consistency can be achieved at the cost of a 7% decrease in time efficiency, I would go for that. Also seems to reduce code complexity, if I understand correctly.

**Comment by rettigl** [comment]
Yes, it would reduce complexity. In a previous commit of the bugfix/binning-tuple branch I had already implemented a mechanism that effectively did that (by making the simplify_binning_arguments function calculating and returning arrays of bin edges for all cases). Then I ran however into the NaN issue which I had not solved yet then. So I discarded that again. 
We can think about going this route once more, but I will do some more benchmarks first.

**Comment by rettigl** [comment]
Some more benchmarks. Indeed, the numby bin_ranges function is substantially faster in most cases, so it is worth finding a realsolution for the problem.

Using Fake data
```
{'numba': {'[100]': [1.6987252216786146,
   1.6677521355450153,
   1.7256830390542746],
  'array_[100]': [5.592638127505779, 5.333082078024745, 5.27593994140625],
  '[1000]': [1.6749149225652218, 1.669021189212799, 1.6788905560970306],
  'array_[1000]': [7.547826029360294, 7.550697688013315, 7.554951421916485],
  '[10000]': [1.7199989426881075, 1.7056615687906742, 1.7257521264255047],
  'array_[10000]': [10.301046442240477,
   10.242325283586979,
   10.261294912546873],
  '[100000]': [1.7546296641230583, 1.7194274198263884, 1.7573092766106129],
  'array_[100000]': [13.327671479433775, 13.294762423262, 13.241678860038519],
  '[1000000]': [2.395800404250622, 2.3401004318147898, 2.39807659573853],
  'array_[1000000]': [29.12009622529149,
   29.938354978337884,
   29.208502577617764],
  '[10000000]': [5.109715972095728, 5.103777650743723, 5.109220668673515],
  'array_[10000000]': [78.84378576092422,
   79.13566474802792,
   78.88594887778163],
  '[100000000]': [23.143115619197488, 22.924620481207967, 22.929617159068584],
  'array_[100000000]': [148.13072645105422,
   147.6098561026156,
   148.1305529344827],
  '[10, 10, 10, 10]': [5.075673183426261,
   4.596144983544946,
   4.608203023672104],
  'array_[10, 10, 10, 10]': [13.095697086304426,
   12.893230756744742,
   12.84233552403748],
  '[50, 50, 50, 50]': [7.9280043467879295,
   7.996025277301669,
   8.102917172014713],
  'array_[50, 50, 50, 50]': [25.98566654138267,
   25.991358008235693,
   26.165666511282325],
  '[100, 100, 100, 100]': [21.919513607397676,
   21.854072347283363,
   21.941189581528306],
  'array_[100, 100, 100, 100]': [46.384209951385856,
   46.531198440119624,
   46.32048335112631]},
 'numpy': {'[100]': [9.131666792556643, 9.258922267705202, 9.157401671633124],
  'array_[100]': [9.134916808456182, 9.06809995509684, 9.13587049022317],
  '[1000]': [11.386119361966848, 11.384984411299229, 11.31133577041328],
  'array_[1000]': [11.371166938915849, 11.350688599050045, 11.327065074816346],
  '[10000]': [13.989046704024076, 13.931806968525052, 13.976310895755887],
  'array_[10000]': [14.005341237410903,
   13.963298097252846,
   13.980862522497773],
  '[100000]': [17.687946442514658, 17.60315467044711, 17.5717801656574],
  'array_[100000]': [17.27234260365367,
   17.228328987956047,
   17.326645243912935],
  '[1000000]': [43.71111403033137, 40.55393264628947, 39.769412307068706],
  'array_[1000000]': [41.919808054342866,
   42.580268785357475,
   42.31772758439183],
  '[10000000]': [195.90568762272596, 195.6512057427317, 195.9465512521565],
  'array_[10000000]': [196.7980314809829,
   197.2245777528733,
   197.66826645098627],
  '[100000000]': [299.98436509259045, 299.7048636749387, 299.8386163134128],
  'array_[100000000]': [312.45018381997943,
   312.218419816345,
   312.0349856466055],
  '[10, 10, 10, 10]': [18.36481824889779,
   18.398745853453875,
   18.42483756504953],
  'array_[10, 10, 10, 10]': [18.068515460938215,
   18.131284087896347,
   18.044690495356917],
  '[50, 50, 50, 50]': [26.732456222176552,
   26.927464704960585,
   26.782072007656097],
  'array_[50, 50, 50, 50]': [26.728361450135708,
   26.91948008351028,
   26.967653203755617],
  '[100, 100, 100, 100]': [59.58390528336167,
   59.896403728052974,
   59.85143642127514],
  'array_[100, 100, 100, 100]': [59.73574538528919,
   59.763350615277886,
   59.80036957189441]}}
```
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/e5b2f210-a758-49a4-b044-0328d3315f42)

Using real data:

```
{'numba': {'[100]': [15.583147123456001,
   15.143555833026767,
   15.174742516130209],
  'array_[100]': [19.810667421668768, 19.78793759085238, 19.66218365356326],
  '[1000]': [15.17918724194169, 15.154532862827182, 15.218390308320522],
  'array_[1000]': [22.115185322239995, 22.226516792550683, 22.0214128959924],
  '[10000]': [15.322058686986566, 15.39234958961606, 15.181360553950071],
  'array_[10000]': [24.094100469723344,
   23.943730095401406,
   24.254597198218107],
  '[100000]': [15.475907802581787, 15.19718998670578, 15.277197239920497],
  'array_[100000]': [26.67471701093018, 26.843847451731563, 26.58928257226944],
  '[1000000]': [15.346630709245801, 14.987116765230894, 14.953460212796926],
  'array_[1000000]': [28.885258216410875,
   28.895677274093032,
   28.71893249079585],
  '[10000000]': [14.726623760536313, 14.571816068142653, 14.468874301761389],
  'array_[10000000]': [31.82599739357829,
   31.936698861420155,
   31.932103913277388],
  '[100000000]': [23.90521669946611, 23.68118892237544, 23.831944158300757],
  'array_[100000000]': [57.91552388109267,
   57.7841891720891,
   57.51736646145582],
  '[10, 10, 10, 10]': [14.665743635967374,
   14.362132869660854,
   14.394754307344556],
  'array_[10, 10, 10, 10]': [22.240646734833717,
   22.105724783614278,
   22.41371532343328],
  '[50, 50, 50, 50]': [17.828197229653597,
   17.339598273858428,
   17.173289941623807],
  'array_[50, 50, 50, 50]': [32.34337836317718,
   32.453712463378906,
   32.3817019239068],
  '[100, 100, 100, 100]': [22.41650689020753,
   22.579455582425,
   22.278247870504856],
  'array_[100, 100, 100, 100]': [43.75690128467977,
   43.77953929826617,
   43.64500958845019]},
 'numpy': {'[100]': [18.25040008686483, 18.484507968649268, 18.41496779024601],
  'array_[100]': [22.589964129030704, 22.517897190526128, 22.60899107530713],
  '[1000]': [19.560966692864895, 19.618389500305057, 19.3791935313493],
  'array_[1000]': [24.741201357915998, 24.604322504252195, 24.81355386786163],
  '[10000]': [20.545997235924006, 20.689207199960947, 20.633256735280156],
  'array_[10000]': [27.0222540050745, 26.801118716597557, 27.12217795290053],
  '[100000]': [22.100820811465383, 22.014182662591338, 22.225591707974672],
  'array_[100000]': [30.289710819721222, 30.29602767340839, 30.37006652727723],
  '[1000000]': [31.304538359865546, 30.581201476976275, 30.131322138011456],
  'array_[1000000]': [45.46893757581711, 45.85862319171429, 45.87023634277284],
  '[10000000]': [81.64935851097107, 81.72993516735733, 81.19441273249686],
  'array_[10000000]': [125.35082096420228,
   126.55637978576124,
   125.41555129922926],
  '[100000000]': [124.73121906258166, 122.7236959952861, 122.8573395833373],
  'array_[100000000]': [216.33020637370646,
   216.79338394850492,
   214.8480192963034],
  '[10, 10, 10, 10]': [20.803610522300005,
   20.890610180795193,
   20.815832758322358],
  'array_[10, 10, 10, 10]': [28.08415401354432,
   28.0222026463598,
   27.896158397197723],
  '[50, 50, 50, 50]': [23.774060774594545,
   23.88482358120382,
   23.854171050712466],
  'array_[50, 50, 50, 50]': [33.46969618089497,
   33.866185177117586,
   33.71881324239075],
  '[100, 100, 100, 100]': [38.47635713592172,
   38.59564808011055,
   38.4897524099797],
  'array_[100, 100, 100, 100]': [56.26634711958468,
   56.01062314026058,
   55.71498975530267]}}
```
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/63004d41-5b64-4431-8f7d-8b65d4da11d1)

**Comment by rettigl** [comment]
More refined tests and a solution to the bugs can be found in PR #120 
These fixes increase very mildly the time for numby binning from bin ranges. Respective banchmarks for comparison are shown here:

Using Fake data (1E9 events, 100 dataframe partitions, 20 cores)
```
{'numba': {'[100]': [1.3824714701622725,
   1.3326191864907742,
   1.320770414546132],
  'array_[100]': [3.35405470430851, 2.9515998605638742, 2.9866851568222046],
  '[1000]': [1.4146253876388073, 1.3566670920699835, 1.3445869572460651],
  'array_[1000]': [4.073723604902625, 4.102077784016728, 4.07465729303658],
  '[10000]': [1.410778583958745, 1.3132344856858253, 1.4086810313165188],
  'array_[10000]': [5.298794306814671, 5.303035782650113, 5.2960352543741465],
  '[100000]': [1.4964418020099401, 1.599214905872941, 1.426206337288022],
  'array_[100000]': [6.964585157111287, 7.179456204175949, 7.078924408182502],
  '[1000000]': [2.2971750013530254, 2.0639795903116465, 2.020409259945154],
  'array_[1000000]': [20.84269709326327,
   21.383612463250756,
   20.802752371877432],
  '[10000000]': [3.789999060332775, 3.8546416219323874, 3.800017664209008],
  'array_[10000000]': [44.431696742773056,
   44.247005604207516,
   44.47994740307331],
  '[100000000]': [14.865165093913674, 14.546345109120011, 14.427957719191909],
  'array_[100000000]': [81.93685097061098,
   82.10859550908208,
   82.02591846510768],
  '[10, 10, 10, 10]': [3.778554553166032,
   3.3695930186659098,
   3.2939494345337152],
  'array_[10, 10, 10, 10]': [7.315976394340396,
   7.091481884941459,
   7.071070296689868],
  '[50, 50, 50, 50]': [6.144146302714944,
   6.0269509721547365,
   6.083842920139432],
  'array_[50, 50, 50, 50]': [14.68567143380642,
   14.745273383334279,
   14.724318318068981],
  '[100, 100, 100, 100]': [14.738606663420796,
   14.84958115965128,
   14.645010240375996],
  'array_[100, 100, 100, 100]': [25.84274333715439,
   25.950486790388823,
   25.907727362588048]},
 'numpy': {'[100]': [7.013337306678295, 6.949560288339853, 6.976224914193153],
  'array_[100]': [6.978837940841913, 6.988193968310952, 6.993342153728008],
  '[1000]': [8.17050121165812, 8.15145430341363, 8.125518847256899],
  'array_[1000]': [8.128429291769862, 8.206220712512732, 8.170202555134892],
  '[10000]': [9.479196917265654, 9.474959133192897, 9.468353198841214],
  'array_[10000]': [9.456732667982578, 9.46544942818582, 9.469119807705283],
  '[100000]': [11.238789416849613, 11.07789352722466, 11.044253632426262],
  'array_[100000]': [10.990378575399518,
   11.026102740317583,
   11.00021849386394],
  '[1000000]': [33.99921986833215, 35.30769258365035, 35.10960731096566],
  'array_[1000000]': [34.91838715597987,
   34.824625791981816,
   34.92711911723018],
  '[10000000]': [108.65771105885506, 108.57992417924106, 108.57706700824201],
  'array_[10000000]': [109.41673639975488,
   109.45627430640161,
   109.5582780316472],
  '[100000000]': [169.91060753911734, 165.5332620535046, 165.58012425340712],
  'array_[100000000]': [171.54198540374637,
   171.5205767787993,
   171.83285589516163],
  '[10, 10, 10, 10]': [11.525769079104066,
   11.39395354129374,
   11.459777414798737],
  'array_[10, 10, 10, 10]': [11.207145009189844,
   11.242569863796234,
   11.274320043623447],
  '[50, 50, 50, 50]': [16.35768086463213,
   16.34447262622416,
   16.461854316294193],
  'array_[50, 50, 50, 50]': [16.376837266609073,
   16.376905784010887,
   16.24531989917159],
  '[100, 100, 100, 100]': [39.81794329546392,
   39.75948998704553,
   39.695327604189515],
  'array_[100, 100, 100, 100]': [39.648558950051665,
   39.636893613263965,
   39.667836021631956]}}
```
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/319ecdd2-e09e-4431-b51b-55dc27cb432b)

Using real data (100 files, 20 cores, local storage):
```
{'numba': {'[100]': [14.733952196314931,
   14.849328998476267,
   15.200421398505569],
  'array_[100]': [16.302211763337255, 16.392615281045437, 16.269374368712306],
  '[1000]': [15.162885587662458, 14.943408912047744, 14.989996191114187],
  'array_[1000]': [17.450123250484467, 17.35203031823039, 17.447106508538127],
  '[10000]': [15.099945977330208, 15.293334448710084, 15.184131419286132],
  'array_[10000]': [18.428123513236642, 18.51760818436742, 18.481907619163394],
  '[100000]': [15.134377336129546, 14.913739308714867, 14.889178587123752],
  'array_[100000]': [19.827850427478552,
   19.82690643146634,
   19.724689975380898],
  '[1000000]': [15.560975816100836, 15.227393181994557, 15.199127046391368],
  'array_[1000000]': [21.04010139964521,
   20.856430204585195,
   20.726763740181923],
  '[10000000]': [15.289620444178581, 15.373270783573389, 15.261911073699594],
  'array_[10000000]': [22.692875657230616,
   22.66524164006114,
   22.920301409438252],
  '[100000000]': [24.66429628431797, 24.19289081916213, 24.5162946972996],
  'array_[100000000]': [37.92821584455669,
   37.43732442520559,
   37.46279442310333],
  '[10, 10, 10, 10]': [15.572990095242858,
   15.705371348187327,
   15.466073589399457],
  'array_[10, 10, 10, 10]': [17.520021840929985,
   17.450484132394195,
   17.43330093473196],
  '[50, 50, 50, 50]': [18.176856637001038,
   17.498917328193784,
   17.86747421324253],
  'array_[50, 50, 50, 50]': [23.098831778392196,
   22.933644196018577,
   23.022386353462934],
  '[100, 100, 100, 100]': [23.596868470311165,
   23.45271508395672,
   23.532816477119923],
  'array_[100, 100, 100, 100]': [30.01842142827809,
   30.245379315689206,
   30.361412147060037]},
 'numpy': {'[100]': [18.524135487154126,
   18.633452463895082,
   18.54890693537891],
  'array_[100]': [18.416799552738667, 18.688915882259607, 18.421326441690326],
  '[1000]': [19.38429730385542, 19.425643460825086, 19.25167927145958],
  'array_[1000]': [19.390602815896273, 19.4795960560441, 19.363884214311838],
  '[10000]': [20.353677704930305, 20.49501223862171, 20.324606770649552],
  'array_[10000]': [20.38441687822342, 20.28081870265305, 20.522769562900066],
  '[100000]': [21.699694084003568, 21.84718623571098, 21.704747308045626],
  'array_[100000]': [21.80632309615612, 21.766659712418914, 21.58275188319385],
  '[1000000]': [32.79759572073817, 29.94868630170822, 31.176664233207703],
  'array_[1000000]': [32.4061451908201, 31.59812618419528, 31.470240524038672],
  '[10000000]': [81.98115051165223, 81.95168690569699, 81.67851208150387],
  'array_[10000000]': [82.82444016076624, 83.1289587225765, 82.66517231613398],
  '[100000000]': [123.24448385834694, 123.04206421598792, 122.82140993326902],
  'array_[100000000]': [129.17909926734865,
   129.42921847477555,
   129.61342964693904],
  '[10, 10, 10, 10]': [21.19589369557798,
   21.27108588628471,
   21.331138975918293],
  'array_[10, 10, 10, 10]': [21.23748799227178,
   21.491274908185005,
   21.39603771455586],
  '[50, 50, 50, 50]': [24.21680619381368,
   24.25677534006536,
   24.10107490979135],
  'array_[50, 50, 50, 50]': [23.94877565652132,
   24.15842237137258,
   23.97690576314926],
  '[100, 100, 100, 100]': [38.34515713900328,
   38.8367358520627,
   38.3656724113971],
  'array_[100, 100, 100, 100]': [38.568836437538266,
   38.54102039709687,
   38.845398887991905]}}
```
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/b16b019c-d64b-451a-bb92-bc5c0cdbf8ee)

Jittered, calibrated data:
```
{'numba': {'[100]': [104.2074211332947,
   104.92128176428378,
   105.14531268924475],
  'array_[100]': [106.09799040108919, 105.55311788059771, 106.01667489111423],
  '[1000]': [105.15801160782576, 104.64784630388021, 105.09364487603307],
  'array_[1000]': [107.57440074160695, 107.37698679603636, 106.6539329700172],
  '[10000]': [104.85559518635273, 104.99284113571048, 104.73894251137972],
  'array_[10000]': [108.21907082758844,
   108.81941968016326,
   109.08836130797863],
  '[100000]': [104.13808564096689, 105.17766770534217, 104.63994487188756],
  'array_[100000]': [109.88417259044945,
   109.30562711507082,
   109.5063901077956],
  '[1000000]': [105.65623974613845, 105.97941259108484, 105.43637097068131],
  'array_[1000000]': [119.21700381673872,
   119.62545640580356,
   118.86234227195382],
  '[10000000]': [110.07981223613024, 110.44180864468217, 110.24265591055155],
  'array_[10000000]': [149.16953836381435,
   149.58566305972636,
   149.7772287223488],
  '[100000000]': [116.25846873410046, 117.0750398747623, 115.89956901222467],
  'array_[100000000]': [184.99378168582916,
   183.66882034391165,
   184.98700869828463],
  '[10, 10, 10, 10]': [105.73681841231883,
   105.83943613804877,
   106.60871942900121],
  'array_[10, 10, 10, 10]': [109.40625412017107,
   108.08974407054484,
   108.07152582146227],
  '[50, 50, 50, 50]': [108.22186208888888,
   108.10032203420997,
   108.01605096273124],
  'array_[50, 50, 50, 50]': [112.63759251870215,
   113.06449491903186,
   114.17500177770853],
  '[100, 100, 100, 100]': [114.0715959109366,
   113.7955147922039,
   113.16979626379907],
  'array_[100, 100, 100, 100]': [120.14550096727908,
   120.28745714016259,
   119.40778942033648]},
 'numpy': {'[100]': [107.32603667490184,
   106.9362608268857,
   108.15561641193926],
  'array_[100]': [107.50039936229587, 108.23094709403813, 107.00013668835163],
  '[1000]': [108.55091469176114, 108.64826738461852, 108.4724689591676],
  'array_[1000]': [108.63636005856097, 107.3910877853632, 107.94656432047486],
  '[10000]': [110.16419153101742, 109.30422992818058, 109.92894489876926],
  'array_[10000]': [110.11387050710618, 109.2186338827014, 110.19697081111372],
  '[100000]': [111.32307031005621, 111.56338098831475, 111.16040220111609],
  'array_[100000]': [110.5581366289407,
   110.81381987966597,
   110.15134275145829],
  '[1000000]': [130.3299080580473, 130.8096560332924, 131.57040036469698],
  'array_[1000000]': [131.25989087671041,
   132.60178405418992,
   131.10198944807053],
  '[10000000]': [211.88258667476475, 211.63313671015203, 211.801554473117],
  'array_[10000000]': [212.65202833153307,
   212.20812638103962,
   213.0549043212086],
  '[100000000]': [279.15031083859503, 275.08830056525767, 268.2997956927866],
  'array_[100000000]': [272.723230458796,
   272.9602278843522,
   273.2770413272083],
  '[10, 10, 10, 10]': [110.84495286643505,
   110.3056786712259,
   111.35439045727253],
  'array_[10, 10, 10, 10]': [110.73045275174081,
   110.60686484351754,
   110.78295120410621],
  '[50, 50, 50, 50]': [114.1368409357965,
   113.94475466758013,
   113.28940320014954],
  'array_[50, 50, 50, 50]': [113.41536262258887,
   113.71775895170867,
   113.61791147850454],
  '[100, 100, 100, 100]': [130.05532558076084,
   129.7188997399062,
   129.02201546542346],
  'array_[100, 100, 100, 100]': [129.66379526257515,
   128.74802346527576,
   129.7020330633968]}}
```
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/d455cf13-b32a-4ed3-9b01-8e32ab55ec8a)

**Comment by steinnymir** [comment]
Im not sure what you are trying to show with the second round of plots. What changed? you sai you fixed the bug in #120? how does this impact performance?

**Comment by rettigl** [comment]
> Im not sure what you are trying to show with the second round of plots. What changed? you sai you fixed the bug in #120? how does this impact performance?

It's not quite visible in the plots, but in the numbers. The numba bins method is a few percent slower with the fix, but still way faster thatn all the other methods. For real data with conversions, the differences are however marginal.
